### ISSUE TITLE
file attributes are lost

### ISSUE THREAD

**Issue by Tcll** [ipfs-shipyard/ipfs-webui#1588]
**Describe the bug**
![Screenshot_2020-08-12_16-57-14](https://user-images.githubusercontent.com/773959/90067310-ed372d00-dcbc-11ea-97b2-ffe41983b8cf.png)
on the left is the original file
on the right is the uploaded file re-downloaded through my client

**To Reproduce**
1. import a file
2. download that file you just imported
3. compare the file attributes

**Expected behavior**
since I'm the one who uploaded, I expect my file attributes to match
if someone else downloads the file however, the file attributes matching should be a hat toss like it is normally (may or may not match)

I use this on occasion to track when I modified said file, for various purposes (always been useful)
(NextCloud does this, but sadly my friend's server has been down for too long and I need a new place)

**Desktop (necessary information):**
 - OS: Xubuntu + Void Linux
 - IPFS Version: (AppImage) https://github.com/ipfs-shipyard/ipfs-webui/commit/7a8f695

**Comment by jessicaschilling** [comment]
@lidel and/or @rafaelramalho19, any thoughts? Thanks!

**Comment by lidel** [comment]
@Tcll  thank you for reporting this. It is a tricky topic with short and long answer :)

### Short answer

WebUI is does not support persisting attributes such as modification time from local filesystem because of the limitations of both IPFS and the web platform. If you need to persist any metadata about your files, you need to do that by other means.

### Longer answer

- `ipfs.add` in js-ipfs supports `mode` and `mtime` (aka unixfsv1.5) as opt-in parameters ([docs](https://github.com/ipfs/js-ipfs/blob/master/docs/core-api/FILES.md#ipfsadddata-options))
  - If you really want to persist modification time, your only option is to use js-ipfs in the commandline and experiment with `jsipfs add --preserve-mtime file`, but read caveat described below
- ipfs-webui does not use this, because
	- go-ipfs does not support this yet
    - it is an educational app, and due to this we want CIDs to be based on file contents alone
      - when `mtime` is persisted as opt-in, the metadata is stored in the DAG itself, so it changes the final CID of the file, and we want different people/machines adding the same file from different locations get the same CID
        - Illustration why its bad for default behavior: if you and your friend add the same file with different modification date, you will get different root CID. If you share CID with someone, that person won't be able to leverage the fact that two people have the file unless they fetch the root block from you. So if you (with your unique root block) go offline, data is not reachable. That is why preserving modification time is opt-in feature with limited use cases.
   - webui  runs in Electron but also in regular browsers, and browsers do not support setting file attributes on downloaded files
     - see [this long discussion about modification time, ending with WONTFIX in Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=178506)
   

----

**TL;DR** WebUI is a web app, and Web platform does not support this. Feel free to reopen if/when that changes.

**Comment by Tcll** [comment]
ah I see, so something that might be done later

btw NextCloud is also used through a web client, which is why I felt it was significant enough to mention ;)
and yes I can confirm, through vivaldi, the attributes are lost, but that just means not using a browser to preserve attributes.

but it's good to know there's at least A method for preserving attributes...
glad the potential for improvement is there :)
